### PR TITLE
Allow EVP_PKEY private key creation without a public component

### LINKED ISSUE #13506: Cannot create DH private key without public value when using OSSL_PARAMs
This is an issue I've run into using the latest 3.0. Previously I've used the now deprecated low-level DH APIs where it was possible to set the private value only using:

`
DH_set0_key(dh, NULL, priv);
`

With the OSSL_PARAMs you'd need to call EVP_PKEY_fromdata with the OSSL_PKEY_PARAM_PRIV_KEY, which ends up calling dh_key_fromdata(). This fails as it also requires the public value as stated in the comment:

```
    /*
     * DH documentation says that a public key must be present if a
     * private key is present.
     * We want to have at least a public key either way, so we end up
     * requiring it unconditionally.
     */
```

For my current application I'm not always able to access the public value so this is a breaking change. Could you consider modifying the behaviour or adding another way to set the private value only?

### PR DESCRIPTION
**Pull request by jon-oracle**
Remove the conditional code which prevents a private key from being created without a public component. Also added a couple of tests.

Fixes #13506

- [x] tests are added or updated

### CODE REVIEW

**Review by paulidale — CHANGES_REQUESTED** [review]

File: test/evp_extra_test.c [hunk 1]
```diff
@@ -1833,7 +1959,133 @@ static int test_decrypt_null_chunks(void)
 }
 #endif /* !defined(OPENSSL_NO_CHACHA) && !defined(OPENSSL_NO_POLY1305) */
 
-#if !defined(OPENSSL_NO_DH) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+#ifndef OPENSSL_NO_DH
+/* Test combinations of private, public, missing and private + public key
```
Comment (slontis) [inline test/evp_extra_test.c:1963]: Format should be the following for multi-liners.
/*
 * Test ...
 * ....
 */
Same applies to the function above..

File: test/evp_extra_test.c [hunk 2]
```diff
@@ -1833,7 +1959,133 @@ static int test_decrypt_null_chunks(void)
...
+    BN_free(p);
+    BN_free(q);
+    BN_free(g);
+    BN_free(pub);
+    BN_free(priv);
+
+    return ret;
+}
+
+#ifndef OPENSSL_NO_DEPRECATED_3_0
```
Comment (slontis) [inline test/evp_extra_test.c:2088]: indent the inner define
```
#ifdef ...
# ifndef ... 
# endif
#endif
```

File: test/evp_extra_test.c [hunk 3]
```diff
@@ -2181,13 +2434,17 @@ int setup_tests(void)
...
+    ADD_TEST(test_DSA_priv_pub);
 #endif
     ADD_TEST(test_RSA_get_set_params);
 #if !defined(OPENSSL_NO_CHACHA) && !defined(OPENSSL_NO_POLY1305)
     ADD_TEST(test_decrypt_null_chunks);
 #endif
-#if !defined(OPENSSL_NO_DH) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+#ifndef OPENSSL_NO_DH
+    ADD_TEST(test_DH_priv_pub);
+#ifndef OPENSSL_NO_DEPRECATED_3_0
```
Comment (slontis) [inline test/evp_extra_test.c:2445]: same here..

File: test/evp_extra_test.c [hunk 4]
```diff
@@ -1833,7 +1959,133 @@ static int test_decrypt_null_chunks(void)
...
+    BIGNUM *p = NULL, *q = NULL, *g = NULL, *pub = NULL, *priv = NULL;
+    EVP_PKEY_CTX *pctx = NULL;
+    EVP_PKEY *pkey = NULL;
+    int ret = 0;
+
+    /*
+     * Setup the parameters for our DH object. For our purposes they don't
+     * have to actually be *valid* parameters. We just need to set something.
+     */
+    if (!TEST_ptr(pctx = EVP_PKEY_CTX_new_from_name(testctx, "DH", NULL))
```
Comment (levitte) [inline test/evp_extra_test.c:1978]: Does this function and the DSA test above differ more than the key type string?

Comment (jon-oracle) [reply test/evp_extra_test.c:1978]: Yes I believe so. Wasn't sure if the behaviour would be the same when I copy/pasted it but it does seem to be. I should probably refactor it...

Comment (levitte) [reply test/evp_extra_test.c]: I would appreciate a merge of the two functions...  I firmly dislike code copying en masse, perhaps it's just me...

Comment (levitte) [reply test/evp_extra_test.c]: I'm assuming a model with a common function, and then individual test functions that call it with different parameters.

Comment (jon-oracle) [reply test/evp_extra_test.c]: Should be fixed in the new commit :)

**Review by levitte — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master as 2c04b34. Thank you for the contribution!

**Pull request closed without merging**